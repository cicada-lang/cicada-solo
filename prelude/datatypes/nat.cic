@module datatype

@datatype Nat {
  zero: Nat
  add1(prev: Nat): Nat
}

Nat.add(x: Nat, y: Nat): Nat
Nat.add(Nat.zero, y) = y
Nat.add(Nat.add1(prev), y) = Nat.add1(Nat.add(prev, y))

Nat.mul(x: Nat, y: Nat): Nat
Nat.mul(Nat.zero, y) = Nat.zero
Nat.mul(Nat.add1(prev), y) = Nat.add(y, Nat.mul(prev, y))

Nat.factorial(x: Nat): Nat
Nat.factorial(Nat.zero) = Nat.add1(Nat.zero)
Nat.factorial(Nat.add1(prev), y) = Nat.mul(x, Nat.factorial(prev))

Nat.one: Nat = Nat.add1(Nat.zero)
Nat.two: Nat = Nat.add1(Nat.one)
Nat.three: Nat = Nat.add1(Nat.two)
Nat.four: Nat = Nat.add1(Nat.three)
Nat.five: Nat = Nat.add1(Nat.four)
Nat.six: Nat = Nat.add1(Nat.five)
Nat.seven: Nat = Nat.add1(Nat.six)
Nat.eight: Nat = Nat.add1(Nat.seven)
Nat.nine: Nat = Nat.add1(Nat.eight)
Nat.ten: Nat = Nat.add1(Nat.nine)

Nat.even_p(x: Nat): Bool
Nat.even_p(Nat.zero) = Bool.true
Nat.even_p(Nat.add1(Nat.zero)) = Bool.false
Nat.even_p(Nat.add1(Nat.add1(prev))) = Nat.even_p(prev)
