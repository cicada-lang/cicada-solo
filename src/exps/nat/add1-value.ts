import { Ctx } from "../../ctx"
import { Core } from "../../core"
import { Value, Subst } from "../../value"
import { readback } from "../../value"
import * as Exps from "../../exps"

export class Add1Value extends Value {
  prev: Value

  constructor(prev: Value) {
    super()
    this.prev = prev
  }

  readback(ctx: Ctx, t: Value): Core | undefined {
    if (t instanceof Exps.NatValue) {
      return new Exps.Add1Core(readback(ctx, t, this.prev))
    }
  }

  unify(subst: Subst, that: Value): Subst {
    if (that instanceof Exps.Add1Value) {
      return subst.unify(this.prev, that.prev)
    } else {
      return Subst.null
    }
  }
}
