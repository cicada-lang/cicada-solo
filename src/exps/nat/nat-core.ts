import { Core, AlphaCtx } from "../../core"
import { Env } from "../../env"
import { Value } from "../../value"
import { Subst } from "../../subst"
import * as Exps from "../../exps"

export class NatCore extends Core {
  evaluate(env: Env): Value {
    return new Exps.NatValue()
  }

  repr(): string {
    return "Nat"
  }

  alpha_repr(ctx: AlphaCtx): string {
    return "Nat"
  }
}
