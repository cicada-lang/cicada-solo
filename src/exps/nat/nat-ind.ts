import { Exp } from "../../exp"
import { Core } from "../../core"
import { Ctx } from "../../ctx"
import { Env } from "../../env"
import { evaluate } from "../../core"
import { check } from "../../exp"
import { Value } from "../../value"
import { Solution } from "../../solution"
import * as Exps from "../../exps"

export class NatInd extends Exp {
  target: Exp
  motive: Exp
  base: Exp
  step: Exp

  constructor(target: Exp, motive: Exp, base: Exp, step: Exp) {
    super()
    this.target = target
    this.motive = motive
    this.base = base
    this.step = step
  }

  free_names(bound_names: Set<string>): Set<string> {
    return new Set([
      ...this.target.free_names(bound_names),
      ...this.motive.free_names(bound_names),
      ...this.base.free_names(bound_names),
      ...this.step.free_names(bound_names),
    ])
  }

  substitute(name: string, exp: Exp): Exp {
    return new NatInd(
      this.target.substitute(name, exp),
      this.motive.substitute(name, exp),
      this.base.substitute(name, exp),
      this.step.substitute(name, exp)
    )
  }

  infer(ctx: Ctx): { t: Value; core: Core } {
    const target_core = check(ctx, this.target, new Exps.NatValue())
    const motive_t = nat_ind_motive_t
    const motive_core = check(ctx, this.motive, motive_t)
    const motive_value = evaluate(ctx.to_env(), motive_core)
    const base_core = check(
      ctx,
      this.base,
      Exps.ApCore.apply(motive_value, new Exps.ZeroValue())
    )
    const step_core = check(ctx, this.step, nat_ind_step_t(motive_value))
    const target_value = evaluate(ctx.to_env(), target_core)

    return {
      t: Exps.ApCore.apply(motive_value, target_value),
      core: new Exps.NatIndCore(target_core, motive_core, base_core, step_core),
    }
  }

  repr(): string {
    const args = [
      this.target.repr(),
      this.motive.repr(),
      this.base.repr(),
      this.step.repr(),
    ].join(", ")

    return `nat_ind(${args})`
  }
}

export const nat_ind_motive_t: Value = evaluate(
  Env.empty,
  new Exps.PiCore("target_nat", new Exps.NatCore(), new Exps.TypeCore())
)

export function nat_ind_step_t(motive: Value): Value {
  return evaluate(
    Env.empty.extend("motive", motive),
    new Exps.PiCore(
      "prev",
      new Exps.NatCore(),
      new Exps.PiCore(
        "almost",
        new Exps.ApCore(new Exps.VarCore("motive"), new Exps.VarCore("prev")),
        new Exps.ApCore(
          new Exps.VarCore("motive"),
          new Exps.Add1Core(new Exps.VarCore("prev"))
        )
      )
    )
  )
}
