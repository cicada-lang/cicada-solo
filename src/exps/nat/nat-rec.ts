import { Exp } from "../../exp"
import { Core } from "../../core"
import { Ctx } from "../../ctx"
import { Env } from "../../env"
import { evaluate } from "../../core"
import { check } from "../../exp"
import { infer } from "../../exp"
import { readback } from "../../value"
import { Value } from "../../value"
import { Subst } from "../../subst"
import * as Exps from "../../exps"
import { nanoid } from "nanoid"

export class NatRec extends Exp {
  target: Exp
  base: Exp
  step: Exp

  constructor(target: Exp, base: Exp, step: Exp) {
    super()
    this.target = target
    this.base = base
    this.step = step
  }

  free_names(bound_names: Set<string>): Set<string> {
    return new Set([
      ...this.target.free_names(bound_names),
      ...this.base.free_names(bound_names),
      ...this.step.free_names(bound_names),
    ])
  }

  subst(name: string, exp: Exp): Exp {
    return new NatRec(
      this.target.subst(name, exp),
      this.base.subst(name, exp),
      this.step.subst(name, exp)
    )
  }

  infer(ctx: Ctx): { t: Value; core: Core } {
    const target_core = check(ctx, this.target, new Exps.NatValue())
    const inferred_base = infer(ctx, this.base)
    const base_t_core = readback(ctx, new Exps.TypeValue(), inferred_base.t)
    const target_name = "nat_rec_target_nat_" + nanoid().toString()
    const motive_core = new Exps.TheCore(
      new Exps.PiCore(target_name, new Exps.TypeCore(), new Exps.NatCore()),
      new Exps.FnCore(target_name, base_t_core)
    )
    const step_core = check(ctx, this.step, nat_ind_step_t(inferred_base.t))

    return {
      t: inferred_base.t,
      core: new Exps.NatIndCore(
        target_core,
        motive_core,
        inferred_base.core,
        step_core
      ),
    }
  }

  repr(): string {
    return `nat_rec(${this.target.repr()}, ${this.base.repr()}, ${this.step.repr()})`
  }
}

function nat_ind_step_t(base_t: Value): Value {
  return evaluate(
    Env.empty.extend("base_t", base_t),
    new Exps.PiCore(
      "prev",
      new Exps.NatCore(),
      new Exps.PiCore(
        "almost",
        new Exps.VarCore("base_t"),
        new Exps.VarCore("base_t")
      )
    )
  )
}
