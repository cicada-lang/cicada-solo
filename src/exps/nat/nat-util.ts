import { Exp, ExpMeta, subst } from "../../exp"
import { Core } from "../../core"
import * as Exps from "../../exps"
import pt from "@cicada-lang/partech"

export function nat_from_number(n: number, meta: ExpMeta): Exp {
  if (n <= 0) {
    return new Exps.Zero(meta)
  } else {
    const almost = nat_from_number(n - 1, meta)
    return new Exps.Add1(almost)
  }
}

export function nat_to_number(exp: Exp | Core): number | undefined {
  if (exp instanceof Exps.ZeroCore || exp instanceof Exps.Zero) {
    return 0
  } else if (exp instanceof Exps.Add1Core || exp instanceof Exps.Add1) {
    const almost = nat_to_number((exp as Exps.Add1Core).prev)
    if (almost !== undefined) {
      return 1 + almost
    } else {
      return undefined
    }
  } else {
    return undefined
  }
}
