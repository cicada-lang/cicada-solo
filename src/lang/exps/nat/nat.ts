import { Core } from "../../core"
import { Ctx } from "../../ctx"
import { Env } from "../../env"
import { Exp, ExpMeta } from "../../exp"
import * as Exps from "../../exps"
import { Value } from "../../value"

export class Nat extends Exp {
  meta: ExpMeta

  constructor(meta: ExpMeta) {
    super()
    this.meta = meta
  }

  free_names(bound_names: Set<string>): Set<string> {
    return new Set()
  }

  subst(name: string, exp: Exp): Exp {
    return this
  }

  evaluate(env: Env): Value {
    return new Exps.NatValue()
  }

  infer(ctx: Ctx): { t: Value; core: Core } {
    return {
      t: new Exps.TypeValue(),
      core: new Exps.NatCore(),
    }
  }

  format(): string {
    return "Nat"
  }
}
