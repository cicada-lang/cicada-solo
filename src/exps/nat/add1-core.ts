import { Core, AlphaCtx } from "../../core"
import { Env } from "../../env"
import { evaluate } from "../../core"
import { Value } from "../../value"
import { Subst } from "../../subst"
import { nat_to_number } from "./nat-util"
import * as Exps from "../../exps"

export class Add1Core extends Core {
  prev: Core

  constructor(prev: Core) {
    super()
    this.prev = prev
  }

  evaluate(env: Env): Value {
    return new Exps.Add1Value(evaluate(env, this.prev))
  }

  repr(): string {
    const n = nat_to_number(this)
    if (n !== undefined) {
      return n.toString()
    } else {
      return `add1(${this.prev.repr()})`
    }
  }

  alpha_repr(ctx: AlphaCtx): string {
    const n = nat_to_number(this)
    if (n !== undefined) {
      return n.toString()
    } else {
      return `add1(${this.prev.alpha_repr(ctx)})`
    }
  }
}
