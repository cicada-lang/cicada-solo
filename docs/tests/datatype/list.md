---
title: List
---

# List

``` cicada
datatype List(E: Type) {
  null: List(E)
  cons(head: E, tail: List(E)): List(E)
}

check! List: (Type) -> Type
check! List(String): Type

check! List.null: List(String)
check! List.null(vague String): List(String)

check! List.cons("a", List.null): List(String)
check! List.cons("a", List.cons("b", List.null)): List(String)
check! List.cons("a", List.cons("b", List.cons("c", List.null))): List(String)
```

# induction List

``` cicada
function induction_list(
  implicit E: Type,
  target: List(E),
  motive: (List(E)) -> Type,
  case_of_null: motive(List.null),
  case_of_cons: (
    head: E, tail: List(E),
    almost: class { tail: motive(tail) },
  ) -> motive(List.cons(head, tail)),
): motive(target) {
  return induction (target) {
    motive
    case null => case_of_null
    case cons(head, tail, almost) => case_of_cons(head, tail, almost)
  }
}
```

# length

``` cicada
import { Nat } from "./nat.md"

function length(implicit E: Type, x: List(E)): Nat {
  return induction (x) {
    case null => Nat.zero
    case cons(_head, _tail, almost) => Nat.add1(almost.tail)
  }
}
```

``` cicada
same_as_chart! (Nat) [
  length(the(List(String), List.cons("a", List.cons("b", List.cons("c", List.null))))),
  Nat.add1(Nat.add1(Nat.add1(Nat.zero))),
]
```

# append

``` cicada
function append(implicit E: Type, x: List(E), y: List(E)): List(E) {
  return induction (x) {
    case null => y
    case cons(head, _tail, almost) => List.cons(head, almost.tail)
  }
}
```

``` cicada
same_as_chart! (List(String)) [
  append(
    the(List(String), List.cons("a", List.cons("b", List.cons("c", List.null)))),
    the(List(String), List.cons("a", List.cons("b", List.cons("c", List.null)))),
  ),
  List.cons("a", List.cons("b", List.cons("c", List.cons("a", List.cons("b", List.cons("c", List.null)))))),
]
```

# reverse

``` cicada
function list_cons_back(implicit E: Type, e: E, x: List(E)): List(E) {
  return induction (x) {
    case null => List.cons(e, List.null)
    case cons(head, _tail, almost) => List.cons(head, almost.tail)
  }
}

function reverse(implicit E: Type, x: List(E)): List(E) {
  return induction (x) {
    case null => List.null
    case cons(head, _tail, almost) => list_cons_back(head, almost.tail)
  }
}
```

``` cicada
same_as_chart! (List(String)) [
  list_cons_back("d", List.cons("a", List.cons("b", List.cons("c", List.null)))),
  List.cons("a", List.cons("b", List.cons("c", List.cons("d", List.null))))
]

same_as_chart! (List(String)) [
  reverse(the(List(String), List.cons("a", List.cons("b", List.cons("c", List.null))))),
  List.cons("c", List.cons("b", List.cons("a", List.null))),
]
```
