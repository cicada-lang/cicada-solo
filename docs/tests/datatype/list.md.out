(_) => MyList(_): (_: Type) -> Type
MyList(Nat): Type
MyList.my_null(returned Nat): MyList(Nat)
MyList.my_null(returned Nat): MyList(Nat)
MyList.my_null(returned String): MyList(String)
MyList.my_null(returned String): MyList(String)
MyList.my_cons(returned Nat, 1, MyList.my_null(returned Nat)): MyList(Nat)
MyList.my_cons(returned Nat, 1, MyList.my_null(returned Nat)): MyList(Nat)
MyList.my_cons(returned Nat, 1, MyList.my_cons(returned Nat, 2, MyList.my_null(returned Nat))): MyList(Nat)
MyList.my_cons(returned Nat, 1, MyList.my_cons(returned Nat, 2, MyList.my_cons(returned Nat, 3, MyList.my_null(returned Nat)))): MyList(Nat)
MyList.my_cons(returned String, "a", MyList.my_null(returned String)): MyList(String)
MyList.my_cons(returned String, "a", MyList.my_cons(returned String, "b", MyList.my_null(returned String))): MyList(String)
MyList.my_cons(returned String, "a", MyList.my_cons(returned String, "b", MyList.my_cons(returned String, "c", MyList.my_null(returned String)))): MyList(String)
refl: Equal(Nat, 3, 3)
refl: Equal(List(Nat), li(1, li(2, li(3, li(4, li(5, li(6, nil)))))), li(1, li(2, li(3, li(4, li(5, li(6, nil)))))))
refl: Equal(List(Nat), li(1, li(2, li(3, li(4, nil)))), li(1, li(2, li(3, li(4, nil)))))
refl: Equal(List(Nat), li(3, li(2, li(1, nil))), li(3, li(2, li(1, nil))))